projects/portal/src/app/models/wallets/metamask/metamask.infrastructure.service.ts

```diff
@@ -0,0 +1,156 @@
+    return signature;
+  }
+
+  private async getPublicKeyFromMetaMask(): Promise<string> {
+    const message = 'Connecting to MetaMask from UnUniFi';
+    const signature = await this.personalSign(message);
+    console.log('signature', signature);
+    const recoveredAddress = ethers.utils.verifyMessage(message, signature);
+    console.log('ethRecoveredAddress', recoveredAddress);
+
+    const digest = this.messageToDigest(message);
+    console.log('digest', digest);
+    const recoveredUncompressedPublicKey = ethers.utils.recoverPublicKey(digest, signature);
+    console.log('ethUncompressedPublicKey', recoveredUncompressedPublicKey);
+    const recoverdAddress2 = ethers.utils.recoverAddress(digest, signature);
+    console.log('ethRecoveredAddress2', recoverdAddress2);
+    const recoveredCompressedPublicKey = ethers.utils.computePublicKey(
+      recoveredUncompressedPublicKey,
+      true,
+    );
+    console.log('ethCompressedPublicKey', recoveredCompressedPublicKey);
+    const publicKeyWithout0x = recoveredCompressedPublicKey.slice(2);
+    return publicKeyWithout0x;
+  }
+
+  private async getEthAddress(): Promise<string> {
+    await detectEthereumProvider({ mustBeMetaMask: true });
+    if (!window.ethereum?.isMetaMask) {
+      throw Error('MetaMask is not installed!');
+    }
+    const provider = new ethers.providers.Web3Provider(window.ethereum);
+    const accounts = await provider.send('eth_requestAccounts', []);
+    if (!accounts?.length) {
+      throw Error('Failed to get MetaMask account!');
+    }
+    const address = accounts[0];
+    console.log(address);
+    return accounts[0];
+  }
+
+  private async getCosmosStoredWalletFromMetaMask(): Promise<StoredWallet> {
+    const ethAddress = await this.getEthAddress();
+    const publicKeyString = await this.getPublicKeyFromMetaMask();
+    const cosmosPublicKey = createCosmosPublicKeyFromString(KeyType.secp256k1, publicKeyString);
+    if (!cosmosPublicKey) {
+      throw Error('Invalid public key!');
+    }
+    const accAddress = cosmosclient.AccAddress.fromPublicKey(cosmosPublicKey);
+    const storedWallet: StoredWallet = {
+      id: ethAddress,
+      type: WalletType.metaMask,
+      key_type: KeyType.ethsecp256k1,
+      public_key: publicKeyString,
+      address: accAddress.toString(),
+    };
+    return storedWallet;
+  }
+
+  async connectWallet(): Promise<StoredWallet | null | undefined> {
+    return await this.getCosmosStoredWalletFromMetaMask();
+  }
+
+  async signTx(
+    txBuilder: cosmosclient.TxBuilder,
+    signerBaseAccount: proto.cosmos.auth.v1beta1.BaseAccount,
+  ): Promise<cosmosclient.TxBuilder> {
+    txBuilder.signDocBytes(signerBaseAccount.account_number);
+    const txJsonString = txBuilder.cosmosJSONStringify();
+    if (!txBuilder) {
+      throw Error('Failed to txBuilder');
+    }
+    const txJson = JSON.parse(txJsonString);
+
+    // fix JSONstringify issue
+    delete txJson.auth_info.signer_infos[0].mode_info.multi;
```

Comment:
by using 
https://github.com/cosmos-client/cosmos-client-ts/pull/71
this PR, you can delete this line.